llvm/lib/CodeGen/WindowsSecureHotPatching.cpp

```diff
@@ -0,0 +1,599 @@
+      ReplacedValue = rewriteGlobalVariablesInConstant(
+          OldConstant, InsertionPoint, GVLoadMap);
+    }
+    // Do not use short-circuiting, here. We need to traverse the whole tree.
+    ReplacedAnyOperands |= ReplacedValue != nullptr;
+    ReplacedValues.push_back(ReplacedValue);
+  }
+
+  // If none of our operands were replaced, then don't rewrite this expression.
+  if (!ReplacedAnyOperands) {
+    return nullptr;
+  }
+
+  // We need to rewrite this expression. Convert this constant expression
+  // to an instruction, then replace any operands as needed.
+  Instruction *NewInst = cast<ConstantExpr>(C)->getAsInstruction();
+  for (unsigned OpIndex = 0; OpIndex < NumOperands; ++OpIndex) {
+    Value *ReplacedValue = ReplacedValues[OpIndex];
+    if (ReplacedValue != nullptr) {
+      NewInst->setOperand(OpIndex, ReplacedValue);
+    }
+  }
+
+  // Insert the new instruction before the reference instruction.
+  IRBuilder<> Builder(InsertionPoint);
+  Builder.Insert(NewInst);
+
+  return NewInst;
+}
+
+// Processes a function that is marked for hot-patching.
+//
+// If a function is marked for hot-patching, we generate an S_HOTPATCHFUNC
+// CodeView debug symbol. Tools that generate hot-patches look for
+// S_HOTPATCHFUNC in final PDBs so that they can find functions that have been
+// hot-patched and so that they can distinguish hot-patched functions from
+// non-hot-patched functions.
+//
+// Also, in functions that are hot-patched, we must indirect all access to
+// (mutable) global variables through a pointer. This pointer may point into the
+// unpatched ("base") binary or may point into the patched image, depending on
+// whether a hot-patch was loaded as a patch or as a base image.  These
+// indirections go through a new global variable, named `__ref_<Foo>` where
+// `<Foo>` is the original symbol name of the global variable.
+//
+// This function handles rewriting accesses to global variables, but the
+// generation of S_HOTPATCHFUNC occurs in
+// CodeViewDebug::emitHotPatchInformation().
+//
+// Returns true if any global variable references were found.
+
+static bool searchConstantExprForGlobalVariables(
+    Value *V, SmallDenseMap<GlobalVariable *, Value *> &GVLoadMap,
+    SmallVector<GlobalVariableUse> &GVUses) {
+
+  SmallVector<Value *, 8> ReplacedOperands;
+
+  if (V->getValueID() == Value::GlobalVariableVal) {
+    GlobalVariable *GV = cast<GlobalVariable>(V);
+    if (globalVariableNeedsRedirect(GV)) {
+      GVLoadMap[GV] = nullptr;
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  if (User *U = dyn_cast<User>(V)) {
+    unsigned NumOperands = U->getNumOperands();
+    bool FoundAny = false;
+    for (unsigned OpIndex = 0; OpIndex < NumOperands; ++OpIndex) {
+      Value *Op = U->getOperand(OpIndex);
+      // Do not use short-circuiting, here. We need to traverse the whole tree.
+      FoundAny |= searchConstantExprForGlobalVariables(Op, GVLoadMap, GVUses);
```

Comment:
Would it make sense to keep track of constants we've walked previously to avoid walking them again?

Reply:
I don't think so.  We already keep track of instructions.  The constant expressions are generally quite small, and we're only processing them in functions that are marked for hot-patching to begin with.